courses/models.py

```diff
@@ -0,0 +1,628 @@
+        title = f"{self.courseware_id} | {self.title}"
+        return title if len(title) <= 100 else title[:97] + "..."
+
+    def clean(self):
+        """
+        If expiration_date is not set:
+        1. If end_date is provided: set expiration_date to default end_date + 90 days.
+        2. If end_date is None, don't do anything.
+
+        Validate that the expiration date is:
+        1. Later than end_date if end_date is set
+        2. Later than start_date if start_date is set
+        """
+        if not self.expiration_date:
+            return
+
+        if self.start_date and self.expiration_date < self.start_date:
+            raise ValidationError("Expiration date must be later than start date.")
+
+        if self.end_date and self.expiration_date < self.end_date:
+            raise ValidationError("Expiration date must be later than end date.")
+
+    def save(
+        self, force_insert=False, force_update=False, using=None, update_fields=None
+    ):
+        """
+        Overriding the save method to inject clean into it
+        """
+        self.clean()
+        super().save(
+            force_insert=force_insert,
+            force_update=force_update,
+            using=using,
+            update_fields=update_fields,
+        )
+
+
+class EnrollmentModel(TimestampedModel, AuditableModel):
+    """Abstract base model for enrollments"""
+
+    class Meta:
+        abstract = True
+
+    user = models.ForeignKey(User, on_delete=models.CASCADE)
+    change_status = models.CharField(
+        choices=ENROLL_CHANGE_STATUS_CHOICES, max_length=20, null=True, blank=True
+    )
+    active = models.BooleanField(
+        default=True,
+        help_text="Indicates whether or not this enrollment should be considered active",
+    )
+
+    objects = ActiveEnrollmentManager()
+    all_objects = models.Manager()
+
+    @classmethod
+    def get_audit_class(cls):
+        raise NotImplementedError
+
+    @classmethod
+    def objects_for_audit(cls):
+        return cls.all_objects
+
+    def to_dict(self):
+        return {
+            **serialize_model_object(self),
+            "username": self.user.username,
+            "full_name": self.user.name,
+            "email": self.user.email,
+        }
+
+    def deactivate_and_save(self, change_status, no_user=False):
+        """Sets an enrollment to inactive, sets the status, and saves"""
+        self.active = False
+        self.change_status = change_status
+        return self.save_and_log(None if no_user else self.user)
+
+    def reactivate_and_save(self, no_user=False):
+        """Sets an enrollment to be active again and saves"""
+        self.active = True
+        self.change_status = None
+        return self.save_and_log(None if no_user else self.user)
+
+
+class CourseRunEnrollment(EnrollmentModel):
+    """
+    Link between User and CourseRun indicating a user's enrollment
+    """
+
+    run = models.ForeignKey("courses.CourseRun", on_delete=models.PROTECT)
```

Comment:
Still have 4 of these `on_delete=models.PROTECT` rules. Any reason for leaving these and not switching to `CASCADE`?